src/cclib/parser/qchemparser.py

```diff
@@ -159,6 +159,82 @@ def after_parsing(self):
                     newbfname = '{}{}'.format(bfcounts[bfname], bfname)
                     self.aonames[bfindex] = '_'.join([atomname, newbfname])
 
+        # Assign the number of core electrons replaced by ECPs.
+        if hasattr(self, 'user_input') and self.user_input.get('rem') is not None:
+            if self.user_input['rem'].get('ecp') is not None:
+                self.coreelectrons = numpy.zeros(self.natom, 'i')
+                elements = [self.table.element[atomno] for atomno in self.atomnos]
+                ecp_is_gen = (self.user_input['rem']['ecp'] == 'gen')
+                if ecp_is_gen:
+                    assert 'ecp' in self.user_input
+                has_iprint = hasattr(self, 'possible_ecps')
+
+                if not ecp_is_gen and not has_iprint:
+                    msg = """ECPs are present, but the number of core \
+electrons isn't printed at all. Rerun with "iprint >= 100", otherwise \
+coreelectrons is incorrect (all zeros)."""
+                    self.logger.warning(msg)
+                elif ecp_is_gen and not has_iprint:
+                    msg = """ECPs are present, but the number of core \
+electrons are only printed for user-provided definitions. Rerun with \
+"iprint >= 100", otherwise coreelectrons may only be partially correct."""
+                    self.logger.warning(msg)
+                    for entry in self.user_input['ecp']:
```

Comment:
How bad is the "partial correctness" here? Do we really want to parse if it's going to be incorrect?

Reply:
Here is an example.

The first case. There is no ECP info in the output, but the number of core electrons can be read from the user input since the only ECP is user-defined (the 10 in `Stevens_ECP   2   10`).

```
$rem
method = hf
basis = gen
ecp = gen
n_frozen_core = fc
ecp_disable_reconstruction = true
$end

$molecule
 0 1
 Al
 H1  Al  r
 H2  Al  r  H1  120.
 F3  Al  r  H1  120.  H2  180.

 r = 1.6
$end

$ecp
 AL
   Stevens_ECP   2   10
   d potential
      1
      1       1.95559     -3.03055
   s-d potential
      2
      0       7.78858      6.04650
      2       1.99025     18.87509
   p-d potential
      2
      0       2.83146      3.29465
      2       1.38479      6.87029
   ****
$end

$basis
 AL
   SP  3  1.00
      0.9011     -0.30377    -0.07929
      0.4495      0.13382     0.16540
      0.1405      0.76037     0.53015
   SP  1  1.00
      0.04874     0.32232     0.47724
   ****
 H
   3-21G
   ****
 F
   3-21G
   ****
$end

```

The second case. Replacing the basis set on fluorine with a built-in ECP means there's no need to enter the number of core electrons, so it can't be read.

```
$rem
method = hf
basis = gen
ecp = gen
n_frozen_core = fc
ecp_disable_reconstruction = true
$end

$molecule
 0 1
 Al
 H1  Al  r
 H2  Al  r  H1  120.
 F3  Al  r  H1  120.  H2  180.

 r = 1.6
$end

$ecp
 AL
   Stevens_ECP   2   10
   d potential
      1
      1       1.95559     -3.03055
   s-d potential
      2
      0       7.78858      6.04650
      2       1.99025     18.87509
   p-d potential
      2
      0       2.83146      3.29465
      2       1.38479      6.87029
   ****
F
sbkjc
   ****
$end

$basis
 AL
   SP  3  1.00
      0.9011     -0.30377    -0.07929
      0.4495      0.13382     0.16540
      0.1405      0.76037     0.53015
   SP  1  1.00
      0.04874     0.32232     0.47724
   ****
 H
   3-21G
   ****
 F
   sbkjc
   ****
$end
```

cclib output:

```
[eric@copper]$ ccget --verbose coreelectrons AlH2F_test.out AlH2F_sbkjc.out
Attempting to read AlH2F_test.out
Identified logfile to be in QChem format
[QChem AlH2F_test.out WARNING] Attribute charge changed value (10 -> 0)
[QChem AlH2F_test.out WARNING] ECPs are present, but the number of core electrons are only printed for user-provided definitions. Rerun with "iprint >= 100", otherwise coreelectrons may only be partially correct.
coreelectrons
[10  0  0  0]
Attempting to read AlH2F_sbkjc.out
Identified logfile to be in QChem format
[QChem AlH2F_sbkjc.out WARNING] Attribute charge changed value (12 -> 0)
[QChem AlH2F_sbkjc.out WARNING] ECPs are present, but the number of core electrons are only printed for user-provided definitions. Rerun with "iprint >= 100", otherwise coreelectrons may only be partially correct.
[QChem AlH2F_sbkjc.out WARNING] Can't calculate coreelectrons for F 4
coreelectrons
[10  0  0  0]
```

The warning for the first output can be safely ignored. I was thinking about putting an additional message that this could be ignored for such cases. For the second output, `coreelectrons[3]` is wrong; from the charge we can tell it should be 2. In the general case, this number can't be guessed.

Because of the charge discrepancy, you could always get the _total_ number of core electrons right, it's just the assignment to individual atoms that might be wrong. But we'd need to dump additional information that is only handled internally right now.

> How bad is the "partial correctness" here?

It's not great, because I wonder if people read the warnings and are careful about this sort of thing. The partial incorrectness may also be an edge case, since I don't know how common it is to mix built-in and user-defined ECPs. I'd think one or the other is more common, but I'm not an ECP user. Maybe @theochemtheo could comment on his typical ECP usage.

> Do we really want to parse if it's going to be incorrect?

Probably not. In that case, I would stop `logfileparser` from automatically creating `coreelectrons`.

Reply:
@berquist I tend to work with organometallics that only contain one metal, so I only use one ECP at a time, but I'm not sure that would hold for all people.

I guess you could get around this if you had a list of Q-chem's built in ECPs for each atom ([See section 8.7.2 of the Q-chem 5.0 manual](http://www.q-chem.com/qchem-website/manual/qchem50_manual/sect-ECP_derivs.html)) but this is kind of a cumbersome way of doing it.

Reply:
It won't happen for now, but I don't think we're going to completely rule out that option. There is interest in being able to parse basis set names from ORCA outputs, which will at minimum require a regex approach. An alternative that may simplify things across parsers would be to keep track of what basis sets/ECPs each package comes with. We're fortunate that this list doesn't change very often.